### PR TITLE
[Refactor] Change to an event-driven radio system.

### LINKED ISSUE #133: [Refactor] Event/loadout driven radio refactor
I recently helped out MC with ACRE by integrating ACRE support into their framework.  Up to them what they do with that from here, but I learned some pretty good stuff from the experience:

* CBA event handlers are easier to use and more flexible than I thought.  Event-driven stuff in SQF is really feasible via these.
* ACRE makes it really easy to detect pre-existing radios in a loadout, and it seems to be possible to configure these radios *immediately* after loadouting.
* Most of the issues with our current ACRE support are self-inflicted.  Scheduled code and the post-loadout radio shuffling cause more issues than they solve.

I propose something similar to what I've just done for MC:
* Take advantage of CBA events to trigger radio configuration after loadouting.  Remove the gearscript -> radio component dependency.
* Use pre-loadout modifications like what we use for uniforms etc, but for radio replacement/addition.  Ship this process out to the radio component by adding a pre-loadout hook.
* Can utilise the new event-based system to reset radio channels upon group swap / SL "send radio channels" action by just sending/hooking the events.

We have outstanding tickets for persisting spatialisation, volume bla bla.  These might be easier to do with the refactor done.

Attaching the modified version of MCFW here, can see the integration I performed in the git commit history.
[MCFW-ACRE.zip](https://github.com/CombinedArmsGaming/CAFE3/files/9946737/MCFW-ACRE.zip)

Addendum: is it possible to do vehicle radio assignment?  at beginning by assigning vehicle to a group?  when a group member boards as driver?  is it possible to make vehicle radios broadcast on speaker?

### PR DESCRIPTION
**Pull request by Bubbus**
### Pull Request Description
**When merged this pull request will:**
- Move to a more expressive radio config system.
- Make radio assignment more reliable.
- Allow SLs to push radio channels to subordinates.
- Add a nicer radio list in the map screen.
- Made creating multi-language roles easier.

### Release Notes
A new radio system has been added which should make it easier and quicker to set up radios for missions.  For more information, check the new `acre_radio_configuration.sqf` file in the `configuration` folder.

In addition, squad leaders can now manage the radio channels for the whole group and automatically push the radio channels onto their squad members - this should allow SLs to solve radio issues without having to ask everyone what net they're on.

Finally, a new radio list has been added which should make it easier to see radio information, and hide it away when not needed.

## IMPORTANT

- [x] Testing has been completed as neccessary, depending on the nature & impact of the changes.
- [x] The Release Notes section below must be filled out appropriately to explain the changes made in this PR. This section will be used in Framework Changelogs.
- [x] If the contribution affects [the wiki](https://github.com/CombinedArmsGaming/CAFE3/wiki), please include your changes in this pull request so the wiki is consistently updated.
- [x] [Contribution Guidelines](https://github.com/CombinedArmsGaming/CAFE3/blob/release/contributing.md) are read, understood and applied.
- [x] Title of this PR uses our standard template `[Descriptor] - Add|Fix|Improve|Change|Make|Remove {changes}`.

Resolves #133

### CODE REVIEW

File: components/gearScript/fn_applyLoadoutModifications.sqf
Comment (Cre8or) [inline components/gearScript/fn_applyLoadoutModifications.sqf:1]: Is this file still needed?

File: components/gearScript/fn_reapplyGear.sqf
Comment (Cre8or) [inline components/gearScript/fn_reapplyGear.sqf:1]: With the addition of this function, we'd have a `fn_applyLoadout` and a `fn_reapplyGear`, with both doing a similar job (especially regarding the custom CBA events being raised). To me, it'd be clearer to have *one* function handling `setUnitLoadout` on units and doing the pre/post-events, rather than copying these snippets into multiple functions. I'll see how many more times this comes up throughout the review, then I'll adjust my comment.

File: components/radio/acre/configuration/acre_preConfiguration.sqf
Comment (Cre8or) [inline components/radio/acre/configuration/acre_preConfiguration.sqf:1]: I see you've separated chunks of code into different files to later join them via includes - presumably to interface with the configuration file. However, imo this is done at the expense of readability: here you're declaring private variables without assigning them to anything or giving them a default value. If I later want to figure out where a variable comes from, I'd have to trace it here throughout the includes, which isn't ideal (especially with how small these files are).

I'd suggest merging the pre/post configuration snippets into the radio component, leaving only the include for the configuration file (so as to not expose component code to the user).

File: components/radio/acre/configuration/macros.hpp
Comment (Cre8or) [inline components/radio/acre/configuration/macros.hpp:1]: What.

Comment (Cre8or) [reply components/radio/acre/configuration/macros.hpp:1]: Ignoring the fact that I have no clue where you pulled this macro value from; where is it even used?

File: components/radio/acre/fn_acre_applyRadioChannelConfig.sqf [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+				_channelName
+			};
+			case "STRING": 
+			{				
+				private _index = _channelList find (_channelName);
+				(_index + 1) max 1
+			};
+			default 
+			{
+				throw format ["Invalid channel name %1 (type %2)", _channelName, typeName _channelName];
```
Comment (Cre8or) [inline components/radio/acre/fn_acre_applyRadioChannelConfig.sqf:55]: Exception handling, in SQF? _Heresy!_

Nah - if it helps, go on.

File: components/radio/acre/fn_acre_initialiseLanguages.sqf
Comment (Cre8or) [inline components/radio/acre/fn_acre_initialiseLanguages.sqf:1]: A mini-function! Cute 👀

File: components/radio/acre/fn_radioListChildrenActions.sqf [hunk 2]
```diff
@@ -0,0 +1,56 @@
+#include "macros.hpp"
+
+params ["_target", "_player", "_params"];
+
+if (isNil "f_var_originalFunction_addRadioActionsToClass") then
+{
+    // Why isn't this easier to do in SQF?  Why does it have to be so brittle?  thanks lads
+    f_var_originalFunction_addRadioActionsToClass = compile preprocessFileLineNumbers "idi\acre\addons\ace_interact\fnc_radioListChildrenActions.sqf";
```
Comment (Cre8or) [inline components/radio/acre/fn_radioListChildrenActions.sqf:8]: Hang on. They've included a function script file, but not defined a function class for it? Surely that's a bug on ACRE's end.

In any case, it'd be cleaner for us to do our own function class for this script - this way we always have it when we need it, and we don't have to manually use compilation commands.

File: components/radio/acre/macros.hpp
Comment (Cre8or) [inline components/radio/acre/macros.hpp:1]: Not a fan of this, but it's not the first occurence, so I'll give it a pass.

File: components/radio/acre/globals.sqf [hunk 3]
```diff
@@ -0,0 +1,12 @@
+f_arr_acre_vestRadios = ["ACRE_PRC343", "ACRE_PRC152", "ACRE_PRC148", "ACRE_BF888S", "ACRE_SEM52SL"];
+f_arr_acre_backpackRadios = ["ACRE_PRC117F", "ACRE_PRC77", "ACRE_SEM70"];
+f_arr_acre_allRadios = f_arr_acre_vestRadios + f_arr_acre_backpackRadios;
+
+// This map is needed because the ACRE team are big nerds.  See comment here:
+// https://github.com/IDI-Systems/acre2/blob/master/extras/examples/mission_setupCompanySignals.sqf
```
Comment (Cre8or) [inline components/radio/acre/globals.sqf:6]: Ah yes, faithful recreation of the real-life reference object, even if it hurts the code. Surely this has never happened to me! 😬

Good workaround, though.

File: components/radio/common/fn_getRadioModInUse.sqf [hunk 4]
```diff
@@ -0,0 +1,17 @@
+params ["_side"];
+
+if (["acre_sys_radio"] call ace_common_fnc_isModLoaded) then 
+{
+	"ACRE"
+}
+else
+{
+	if (["task_force_radio"] call ace_common_fnc_isModLoaded) then
```
Comment (Cre8or) [inline components/radio/common/fn_getRadioModInUse.sqf:9]: I gotta ask. We don't normally handle compatibility with TFAR. In fact, the functions above (`fn_getRadioList`, `fn_getRadioChannelsForUnit`, etc) don't return anything when ACRE isn't loaded (and ignore TFAR altogether). Why do it here?

Asking for consistency; I don't mind supporting TFAR but I fear this might be a massive undertaking at this point.

File: components/radio/common/radio_configuration.sqf
Comment (Cre8or) [inline components/radio/common/radio_configuration.sqf:1]: Refer to my comment about merging the pre/post files into here.

File: components/radio/globals.sqf
Comment (Cre8or) [inline components/radio/globals.sqf:1]: Are there plans to support other radio systems? If not, do we need this file?

File: components/radio/macros.hpp
Comment (Cre8or) [inline components/radio/macros.hpp:1]: Not gonna lie, I lost track of which one is the original `macros.hpp`.

🏓

**Review by Cre8or — CHANGES_REQUESTED** [review]

File: components/gearScript/fn_reapplyGear.sqf
Comment (Cre8or) [reply components/gearScript/fn_reapplyGear.sqf:1]: So turns out, it's just these two files. Feel free to ignore my original comment.

File: components/gearScript/fn_applyLoadoutModifications.sqf
Comment (Bubbus) [reply components/gearScript/fn_applyLoadoutModifications.sqf:1]: Nah I guess not, it's been outmoded by events now, nice catch.

File: components/gearScript/fn_reapplyGear.sqf
Comment (Bubbus) [reply components/gearScript/fn_reapplyGear.sqf:1]: Fair point tho 👍

File: components/radio/acre/configuration/acre_preConfiguration.sqf
Comment (Bubbus) [reply components/radio/acre/configuration/acre_preConfiguration.sqf:1]: The config files are intended for user readability first - it sucks more as a developer but if it's more readable to MMs then it's the right way imo.  That said, I'll collapse this whole deal down a bit because yeah it's a bit extra rn.

File: components/radio/acre/configuration/macros.hpp
Comment (Bubbus) [reply components/radio/acre/configuration/macros.hpp:1]: This value is chosen to be vanishingly unlikely as a group name or role name - it's used to represent "all" for the purposes of `fn_acre_giveRadioToAllInGroup` etc for dictionary key formation.  You can see an example of usage in `components\radio\acre\configuration\fn_acre_getRadiosForRoleInGroup.sqf`.

File: components/radio/acre/macros.hpp
Comment (Bubbus) [reply components/radio/acre/macros.hpp:1]: Yeah I've got these throughout the whole damn framework, I got annoyed one day when I moved a file with an `#include '..\..\macros.hpp'` in it and it crashed my game, so now every folder has a `macros.hpp`!  its good

File: components/radio/acre/configuration/macros.hpp
Comment (Cre8or) [reply components/radio/acre/configuration/macros.hpp:1]: Hmm. Not a fan of the "ALL" macro, that'll eventually pop up elsewhere and cause script compilation errors (and headaches), but I see your reasoning.

That said though, a human-readable value would be much better than a "garbage" sentinel value. ACE/CBA typically handles this with the GVAR/EGVAR macro functions, which essentially add a prefix to whatever value is passed. Something similar (minus the macro function) could work here too - maybe "CAFE3_SPECIAL_VALUE:ALL"? It's still unlikely to be used (variable names can't contain colons), and its purpose is easier to deduce.

File: components/radio/acre/macros.hpp
Comment (Cre8or) [reply components/radio/acre/macros.hpp:1]: *Is it, though?*

File: components/radio/acre/fn_radioListChildrenActions.sqf (hunk 2 above)
Comment (Bubbus) [reply components/radio/acre/fn_radioListChildrenActions.sqf:8]: Wasn't really clear about the cleanest solution to this - the objective is to have the CAFE radio actions in the same list as the default ones, so you can find all the radio related stuff in the same place.  After digging into it, found a script with a private-variable array of user actions created by this function's counterpart.  Only way I know to achieve the goal is a 'function override' like this, and it sucks.

File: components/radio/common/fn_getRadioModInUse.sqf (hunk 4 above)
Comment (Bubbus) [reply components/radio/common/fn_getRadioModInUse.sqf:9]: I don't really think we'll be switching to TFAR at this point but structuring the radio component as "mod agnostic" on the outside givs us leeway in case we do - the ACRE2 devs might fuck up or move on or something and then we can re-engineer the inside and leave outside references to `common` in place.

File: components/radio/globals.sqf
Comment (Bubbus) [reply components/radio/globals.sqf:1]: *Kinda*, talked about it in the `fn_getRadioModInUse.sqf` comment.

File: components/radio/macros.hpp
Comment (Bubbus) [reply components/radio/macros.hpp:1]: ehehe

File: components/radio/acre/configuration/macros.hpp
Comment (Bubbus) [reply components/radio/acre/configuration/macros.hpp:1]: Yeah anyone who names a squad that is intentionally looking for trouble.  Which the framework doesn't really defend against, but we as people sure can 👀.

File: components/radio/common/fn_getRadioModInUse.sqf (hunk 4 above)
Comment (Cre8or) [reply components/radio/common/fn_getRadioModInUse.sqf:9]: Right, so more for abstraction - that's good then. 👍

File: components/radio/acre/fn_radioListChildrenActions.sqf (hunk 2 above)
Comment (Cre8or) [reply components/radio/acre/fn_radioListChildrenActions.sqf:8]: Ah, I see what you mean. Two parts about this then:

- Compiling the function on the fly strikes me as odd. Normally, addon-defined functions are either available, or they're not (depending on compilation time, not just on mod presence alone). Having to do it ourselves bothers me (and I think you too, judging by the comments), but we could at least move it into a separate init file (or, again, define our own class so it's always there).
- Fetching the child actions by running the base function first is... *not ideal*, admittedly, but it gets the job done. I've looked at ACE's interaction framework, and I'm still not entirely sure what options we have. There *is* an `_insertChildren` argument, but I don't know if it does what we need. Could be worth investigating? In any case, your current approach does the job.

Comment (Bubbus) [reply components/radio/acre/fn_radioListChildrenActions.sqf:8]: I'll admit I'm not entirely grasping what you're suggesting here but I'll take some help if you're willing and it's worth the effort.

File: components/radio/acre/macros.hpp
Comment (Bubbus) [reply components/radio/acre/macros.hpp:1]: its good tho

File: components/radio/acre/fn_radioListChildrenActions.sqf (hunk 2 above)
Comment (Cre8or) [reply components/radio/acre/fn_radioListChildrenActions.sqf:8]: Upon re-reading my comment, I now realise I shouldn't be doing code reviews this late. I'll get back to you on this once I've had some time to sleep, lol

**Comment by Cre8or** [comment]
As discussed, I'd modify the ACRE override to use two distinct functions: your override, and the original function (compiled in the traditional way in `CfgFunctions`. but using a soft-dependency). The functions config would then become

```cpp
#if __has_include("\idi\acre\addons\main\stringtable.xml")
    class acre_ace_interact
    {
        class overrides
        {
            file = "components\radio\acre";
            class radioListChildrenActions{};
            class radioListChildrenActions_original
            {
                file = "idi\acre\addons\ace_interact\fnc_radioListChildrenActions.sqf";
            };
        }
    }
#endif
```

Then we can omit lines 5-12 of your override:

```sqf
if (isNil "f_var_originalFunction_addRadioActionsToClass") then
{
    // Why isn't this easier to do in SQF?  Why does it have to be so brittle?  thanks lads
    f_var_originalFunction_addRadioActionsToClass = compile preprocessFileLineNumbers "idi\acre\addons\ace_interact\fnc_radioListChildrenActions.sqf";
};


private _children = _this call f_var_originalFunction_addRadioActionsToClass;
```

and just do:

```sqf
private _children = _this call acre_ace_interact_fnc_radioListChildrenActions_original;
```

File: gearscript_macros.hpp
Comment (Cre8or) [inline gearscript_macros.hpp:1]: That function is now gone - good catch.

**Review by Cre8or — APPROVED** [review]
Everything's resolved as far as I can see.

🚢
